man/man1/fgviewer.1
.\" Copyright (C) 2017 Alessandro Menti
.\"
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
.\" Or try here: http://www.fsf.org/copyleft/gpl.html
.\"
.TH FGVIEWER 1 2017-06-25 FlightGear "FlightGear man pages"
.SH NAME
fgviewer \- An alternative FlightGear viewer application
.SH SYNOPSIS
\fBfgviewer\fR [\fB\-\-fg\-root\fR \fIrootdir\fR]
[\fB\-\-fg\-scenery\fR \fIscenerydir\fR] [\fB\-\-config\fR \fIconfigfile\fR]
[\fB\-\-prop\fR \fIproperty\fR \fIvalue\fR] [\fB\-\-renderer\fR \fIrend\fR]
[\fB\-\-hla\fR] [\fB\-\-federation\fR \fIfed\fR]
.SH DESCRIPTION
.B fgviewer
is a viewer application that allows you to examine FlightGear scenery and
objects without starting
.BR fgfs (1).
.SH OPTIONS
.TP
\fB\-\-fg\-root\fR \fIrootdir\fR
Set the FlightGear data root directory (\fB$FG_ROOT\fR) to \fIrootdir\fR. If
this option is not set,
.B fgviewer
uses the path specified in the
.B FG_ROOT
environment variable or, absent that, the FlightGear data directory chosen at
the time the program was compiled.
.TP
\fB\-\-fg\-scenery\fR \fIscenerydir\fR
Set the FlightGear scenery directory (\fB$FG_SCENERY\fR) to \fIscenerydir\fR.
If this option is not set,
.B fgviewer
uses the path specified in the
.B FG_SCENERY
environment variable or, absent that, the
.B Scenery
subdirectory inside \fB$FG_ROOT\fR.
.TP
\fB\-\-config\fR \fIconfigfile\fR
Read configuration directives from \fIconfigfile\fR.
.TP
\fB\-\-prop\fR \fIproperty\fR \fIvalue\fR
Set the property named \fIproperty\fR to \fIvalue\fR.
.TP
\fB\-\-renderer\fR \fIrend\fR
Use the renderer named \fIrend\fR to render the scenery. This option is
currently not supported.
.TP
\fB\-\-hla\fR
Enable HLA federation.
.TP
\fB\-\-federation\fR \fIfed\fR
Enable a custom federation named \fIfed\fR.
.SH "EXIT STATUS"
.B fgviewer
exits with
.B EXIT_SUCCESS
on success, with
.B EXIT_FAILURE
if the syntax of the command-line arguments is wrong or an error occurs.
.SH ENVIRONMENT
.IP "\fBFG_ROOT\fR" 4
If
.B FG_ROOT
is set and
.B \-\-fg\-root
is not set, it specifies the root data directory to use.
.IP "\fBFG_SCENERY\fR" 4
If
.B FG_SCENERY
is set and
.B \-\-fg\-scenery
is not set, it specifies the scenery directory to use.
.SH FILES
.IP "\fI$FG_ROOT/defaults.xml\fR" 4
Default preferences file loaded on startup.
.SH "SEE ALSO"
.BR fgfs (1)


NEWS
Version 2.0.0 - February 25, 2010

  FlightGear 2.0.0. reflects the maturation of the OpenSceneGraph port
  that started with the previous 1.9.0 release. In addition to many
  internal code improvements, FlightGear 2.0.0. marks the introduction
  of many new exciting improvements in the graphics and sound system,
  as well as improved usability of key features, and improved behavior
  of exsisting features. Highlights of this new version include:
  Dramatic new 3D clouds, dramatic lighting conditions, improved
  support for custom scenery, and many many new and detailed aircraft
  models.

  Sound 
  * Complete overhaul of the sound code
  * doppler effects
  * distance attenuation
  * 3D positional sound sources
  * assignment of sound sources to external objects (i.e. AI controlled
    aircraft)
  * User selection of the sound device

  Visual Effects
  * Use of Shaders for dynamic textures
  * Use of Effects files
  * Improved 3D clouds
  * Color changes based on humidity and other weather effects allow for very
    dramatic lighting conditions
  * Dynamic water textures
  * Text animation based on OSGText

  Usability
  * Allow screenshots in more common file formats
  * User selectable sound device
  * More intuitive selection of the weather settings through the GUI and/or
    commandline

  Infrastructure
  * Airport geometry data can be read from the scenery, allowing for more
    flexible regeneration of terrain tiles

  Internals
  * Improved efficiency of the property tree
  * A more efficient ground cache
  * Many improvements to the route management code
  * Removed many compiler warnings
  * More realistic atmosphere model

  Behavior
  * More realistic ILS behavior
  * Autopilot improvements
  * A generic autobrake function
  * Winds over mountainous areas cause up- and downdrafts that can be used
    for gliding
  * More realistic behavior of the route manager
  * Wild fires, which can be extinguished by firefighter aircraft operating
    across the multplayer server
  * Navaid frequencies and radials can be transmitted to Atlas

  Utilities
  * A python script to visualize Yasim configuration files in Blender

  AI
  * Allow traffic departing and arriving at the same airport
  * Add Ground Vehicles - including automobiles, trucks, articulated trucks,
    trains (including high speed trains)
  * ATC interactions between AI aircraft and ground controllers
  * Performance characteristics of AI aircraft can be specified in a
    performance database
  * Push-back vehicles are available for a selected number of aircraft
  * Add escorts for AI carrier - frigates, guided missile cruiser, amphibious
    warfare ships now make up the Vinson Battle Group
  * Improved radar functionality - now detects AI escorts etc.
  * AI objects are now solid (i.e. users can collide with them)
  * Some preliminary support for SID/STAR procedures for AI aircraft

Version 1.9.0
* December 20, 2008 (source code snapshot release)


New in 0.9.10
* April 5, 2006

* Many improvements to the multi-player mode.  Other airplanes are now
  updated smoothly.  There is a really nifty google maps based moving
  map that anyone can use to track all online activity.
* AI planes have improved ground following and are smarter about
  selecting the active runway.
* Include the ability to send ATC and AI aircraft communication to a
  local festival server to be converted to spoken speech output.
* Added a detailed model of the Honeywell MK VIII EGPWS ground
  proximity warning system.
* Complete overhaul/redo of terrain textures based on USGS satellite imagery.
* Added background ATC chatter.
* Updated/improved rain cone rendering.
* Updated manual.

* Upgraded to JSBSim v2.0
* Fix a discrepancy in sign of twist and incidence values in YASim
  between code and documentation.
* Updates to YASim for more fine grained propeller modeling.
* UFO will now honor elevator trim value so that it will respond to
  the autopilot altitude hold module.
* UFO can adjust it's speed range so you can use it for super high
  speed world exploration, or extremly fine grained precise positioning.
* Much work and progress with carrier interaction and launch/recovery
  modeling.

* Add a feature that allows the end user to click any where on the
  terrain.  The system will compute a ground intersection and report
  back the lon, lat, and elevation of the click point.  The click
  point is reported in the property system so it can be read by just
  about anything including external programs.  Later we added the
  ability to place an object at the click point as well.
* Add a feature to play one-off wav files.  These are qeued and played
  sequentially.  This was then leveraged to build an ATC background
  chatter module in nasal that can be turned on/off at run time.  This
  could also be leveraged for tutorials or other interactive type
  scripts to play prerecorded messages at specific times or under
  specific conditions..
* Allow clock to report local time as well as universal.
* Add a model of a KLN89 GPS.
* Allow nav radios to be slaved to the gps.
* Allow enabling/disabling of menu entries.
* Replace old hardcoded frame rate display with a much more flexible
  nasal/gui version.
* Add a listener function to properties so that nasal code can be
  triggered whenever a property changes.  This is more efficient than
  polling.
* Add a save-on-exit option to save user preferences.
* Add support for point sprites.
* Add support for emissive layers (such as lights) on 2d panels.
* Updated gui styles.
* Updated airport and navaid data.
* Added a new Nasal/XML based tutorial system.

* Fix a bug in the metar (real weather) threading that could lead to
  long pauses and even crashes.
* Various compiler/platform fixes.
* Fix various valgrind complaints.
* Better use of exceptions in places.
* Fix the 2d compass texture handling.
* Many tweaks and improvements to our keyboard input handling system.

src/Viewer/fgviewer.cxx
#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include <iostream>
#include <cstdlib>

#include <osg/ArgumentParser>
#include <osg/Fog>
#include <osgDB/ReadFile>
#include <osgDB/Registry>
#include <osgDB/WriteFile>
#include <osgViewer/Renderer>
#include <osgViewer/Viewer>
#include <osgViewer/ViewerEventHandlers>
#include <osgGA/KeySwitchMatrixManipulator>
#include <osgGA/TrackballManipulator>
#include <osgGA/FlightManipulator>
#include <osgGA/DriveManipulator>
#include <osgGA/TerrainManipulator>
#include <osgGA/StateSetManipulator>

#include <simgear/props/props.hxx>
#include <simgear/props/props_io.hxx>
#include <simgear/misc/sg_path.hxx>
#include <simgear/scene/material/EffectCullVisitor.hxx>
#include <simgear/scene/material/matlib.hxx>
#include <simgear/scene/util/SGReaderWriterOptions.hxx>
#include <simgear/scene/tgdb/userdata.hxx>
#include <simgear/scene/model/ModelRegistry.hxx>
#include <simgear/scene/model/modellib.hxx>
#include <simgear/structure/exception.hxx>

#include <Scenery/scenery.hxx>

#include <Main/fg_props.hxx>
#include <Main/globals.hxx>
#include <Main/options.hxx>
#include <Main/fg_init.hxx>

class GraphDumpHandler : public  osgGA::GUIEventHandler
{
public:
    GraphDumpHandler() : _keyDump('d') {}
    void setKeyDump(int key) { _keyDump = key; }
    int getKeyDump() const { return _keyDump; }
    bool handle(const osgGA::GUIEventAdapter& ea, osgGA::GUIActionAdapter& aa);

    /** Get the keyboard and mouse usage of this manipulator.*/
    virtual void getUsage(osg::ApplicationUsage& usage) const;
protected:
    int _keyDump;
};

static void dumpOut(osg::Node* node)
{
    char filename[24];
    static int count = 1;

    while (count < 1000) {
        FILE *fp;
        snprintf(filename, 24, "fgviewer-%03d.osg", count++);
        if ( (fp = fopen(filename, "r")) == NULL )
            break;
        fclose(fp);
    }

    if (osgDB::writeNodeFile(*node, filename))
        std::cerr << "Entire scene graph saved to \"" << filename << "\".\n";
    else
        std::cerr << "Failed to save to \"" << filename << "\".\n";
}

bool GraphDumpHandler::handle(const osgGA::GUIEventAdapter& ea,
                              osgGA::GUIActionAdapter& aa)
{
    osgViewer::View* view = dynamic_cast<osgViewer::View*>(&aa);
    if (!view)
        return false;
    if (ea.getHandled())
        return false;
    switch(ea.getEventType()) {
    case osgGA::GUIEventAdapter::KEYUP:
        if (ea.getKey() == _keyDump) {
            dumpOut(view->getScene()->getSceneData());
            return true;
        }
        break;
    default:
        return false;
    }
    return false;
}

void GraphDumpHandler::getUsage(osg::ApplicationUsage& usage) const
{
    std::ostringstream ostr;
    ostr << char(_keyDump);
            usage.addKeyboardMouseBinding(ostr.str(),
                                          "Dump scene graph to file");
}

int
fgviewerMain(int argc, char** argv)
{

    sgUserDataInit(0);

    // use an ArgumentParser object to manage the program arguments.
    osg::ArgumentParser arguments(&argc, argv);

    // construct the viewer.
    osgViewer::Viewer viewer(arguments);
    osg::Camera* camera = viewer.getCamera();
    osgViewer::Renderer* renderer
        = static_cast<osgViewer::Renderer*>(camera->getRenderer());
    for (int i = 0; i < 2; ++i) {
        osgUtil::SceneView* sceneView = renderer->getSceneView(i);
        sceneView->setCullVisitor(new simgear::EffectCullVisitor);
    }
    // Shaders expect valid fog
    osg::StateSet* cameraSS = camera->getOrCreateStateSet();
    osg::Fog* fog = new osg::Fog;
    fog->setMode(osg::Fog::EXP2);
    fog->setColor(osg::Vec4(1.0, 1.0, 1.0, 1.0));
    fog->setDensity(.0000001);
    cameraSS->setAttributeAndModes(fog);
    // ... for some reason, get rid of that FIXME!
    viewer.setThreadingModel(osgViewer::Viewer::SingleThreaded);

    // set up the camera manipulators.
    osgGA::KeySwitchMatrixManipulator* keyswitchManipulator;
    keyswitchManipulator = new osgGA::KeySwitchMatrixManipulator;

    keyswitchManipulator->addMatrixManipulator('1', "Trackball",
                                               new osgGA::TrackballManipulator);
    keyswitchManipulator->addMatrixManipulator('2', "Flight",
                                               new osgGA::FlightManipulator);
    keyswitchManipulator->addMatrixManipulator('3', "Drive",
                                               new osgGA::DriveManipulator);
    keyswitchManipulator->addMatrixManipulator('4', "Terrain",
                                               new osgGA::TerrainManipulator);
    viewer.setCameraManipulator(keyswitchManipulator);

    // Usefull stats
    viewer.addEventHandler(new osgViewer::HelpHandler);
    viewer.addEventHandler(new osgViewer::StatsHandler);
    viewer.addEventHandler( new osgGA::StateSetManipulator(viewer.getCamera()->getOrCreateStateSet()) );
    // Same FIXME ...
    // viewer.addEventHandler(new osgViewer::ThreadingHandler);
    viewer.addEventHandler(new osgViewer::LODScaleHandler);
    viewer.addEventHandler(new osgViewer::ScreenCaptureHandler);

    viewer.addEventHandler(new GraphDumpHandler);

    // Extract files to load from arguments now; this way fgInitConfig
    // won't choke on them.
    string_list dataFiles;
    for (int i = arguments.argc() - 1; i >= 0; --i) {
        if (arguments.isOption(i)) {
            break;
        } else {
            dataFiles.insert(dataFiles.begin(), arguments[i]);
            arguments.remove(i);
        }
    }

    // A subset of full flightgear initialization.
    // Allocate global data structures.  This needs to happen before
    // we parse command line options

    globals = new FGGlobals;

    int configResult = fgInitConfig(arguments.argc(), arguments.argv(), false);
    if (configResult == flightgear::FG_OPTIONS_ERROR) {
        return EXIT_FAILURE;
    } else if (configResult == flightgear::FG_OPTIONS_EXIT) {
        return EXIT_SUCCESS;
    }

    osgDB::FilePathList filePathList
        = osgDB::Registry::instance()->getDataFilePathList();
    filePathList.push_back(globals->get_fg_root().local8BitStr());

    const PathList& path_list = globals->get_fg_scenery();
    for (unsigned i = 0; i < path_list.size(); ++i) {
        filePathList.push_back(path_list[i].local8BitStr());
    }

    globals->set_matlib( new SGMaterialLib );
    simgear::SGModelLib::init(globals->get_fg_root().local8BitStr(), globals->get_props());

    // Initialize the material property subsystem.

    SGPath mpath( globals->get_fg_root() );
    mpath.append( fgGetString("/sim/rendering/materials-file") );
    if ( ! globals->get_matlib()->load(globals->get_fg_root().local8BitStr(),
                                       mpath.local8BitStr(),
            globals->get_props()) ) {
        throw sg_io_exception("Error loading materials file", mpath);
    }

    // The file path list must be set in the registry.
    osgDB::Registry::instance()->getDataFilePathList() = filePathList;

    simgear::SGReaderWriterOptions* options = new simgear::SGReaderWriterOptions;
    options->getDatabasePathList() = filePathList;
    options->setMaterialLib(globals->get_matlib());
    options->setPropertyNode(globals->get_props());
    options->setPluginStringData("SimGear::PREVIEW", "ON");

    FGScenery* scenery = globals->add_new_subsystem<FGScenery>();
    scenery->init();
    scenery->bind();

    // read the scene from the list of file specified command line args.
    osg::ref_ptr<osg::Node> loadedModel;
    loadedModel = osgDB::readNodeFiles(dataFiles, options);

    // if no model has been successfully loaded report failure.
    if (!loadedModel.valid()) {
        std::cerr << arguments.getApplicationName()
                  << ": No data loaded" << std::endl;
        return EXIT_FAILURE;
    }

    // pass the loaded scene graph to the viewer.
    viewer.setSceneData(loadedModel.get());

    int result = viewer.run();
    
    // clear cache now, since it contains SimGear objects. Otherwise SG_LOG
    // calls during shutdown will cause crashes.
    osgDB::Registry::instance()->clearObjectCache();
    
    return result;
}
